### PR TITLE
Add `sdk env install` subcommand

### LINKED ISSUE #784: Feature: Add `sdk env install` subcommand
- [x] I have read the [CONTRIBUTING guidelines](CONTRIBUTING.md)

**Feature request**
As discussed some time ago in the Slack `cli-development` channel, it would be nice to have an additional subcommand for `sdk env` which installs the candidates contained in `.sdkmanrc`.

### PR DESCRIPTION
**Pull request by helpermethod**
Fixes #784.

### CODE REVIEW

**Comment by helpermethod** [comment]
Hi @marc0der,

in the context of this PR I was thinking about this piece of code

```java
		if ! __sdkman_matches_candidate_format "$line"; then
			__sdkman_echo_red "Invalid candidate format!"
			echo ""
			__sdkman_echo_yellow "Expected 'candidate=version' but found '$line'"

			return 1
		fi
```

I wonder if it actually brings much additional value. The format for `.sdkmanrc` is so simple that probably only very few people actually run into issues. Which makes this pretty much dead code. What do you think?

File: gradle/wrapper/gradle-wrapper.properties [hunk 1]
```diff
@@ -1,5 +1,5 @@
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-6.3-bin.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-6.3-all.zip
```
Comment (marc0der) [inline gradle/wrapper/gradle-wrapper.properties:3]: This causes big downloads on CI, mind keeping this as it was?

Comment (helpermethod) [reply gradle/wrapper/gradle-wrapper.properties:3]: Ah, damn, that was definitely not intended! Will remove this ASAP.

**Comment by helpermethod** [comment]
There is now a "higher-order function" __`sdkman_env_each_line()` because `install` and `env` does almost the same. The naming could probably be improved 😄 .

File: src/main/bash/sdkman-env.sh [hunk 2]
```diff
@@ -73,16 +52,36 @@ function __sdkman_generate_sdkmanrc() {
 	__sdkman_echo_green "$sdkmanrc created."
 }
 
-function __sdkman_is_blank_line() {
-	[[ -z "$1" ]]
-}
+function __sdkman_env_each_line() {
```
Comment (marc0der) [inline src/main/bash/sdkman-env.sh:55]: Should this be called something like `__sdkman_env_each_candidate` instead? Maybe I misunderstood, but won't this function process the entire file for either the `install` or `use` subcommands? The loop actual looping of individual lines happening from line 67 onward.

Comment (helpermethod) [reply src/main/bash/sdkman-env.sh:55]: Yes exactly, thanks!

**Comment by helpermethod** [comment]
There seem to be some problems with the `install` command.

image

**Comment by helpermethod** [comment]
Similar behaviour if the first candidate is made the new default.
image

**Comment by marc0der** [comment]
Here come the complications :smile:

In other words, when `sdk env install` is invoked:
* selfupdate functionality should be omitted
* always answer _no_ to setting new version as default

Is that what you had in mind?

**Comment by helpermethod** [comment]
You were faster than me ;). Damn, was pretty sure this would be a small change 😕 . Any invocation of `read` is problematic because it consumes the next line.

Maybe I could read the whole file in one operation and then just iterate over the (in-memory) list? I guess that should get rid of most of the problems.

image

**Comment by marc0der** [comment]
It feels for me like the approach taken with line-by-line reading is the right one.

What would happen if we set `sdkman_auto_answer` to true temporarily in that function, and make the default `n`? The selfupdate issue seems like a separate concern that might result in some deeper surgery when we decouple it, but there we could live with the know issue as this doesn't happen too often.

**Comment by helpermethod** [comment]
Hi,

lines would still be read line by line

```sh
	while IFS= read -r line || [[ -n "$line" ]]; do
		lines+=("$line")
	done

	for line in "${lines[@]}"; do
		line="${line/\#*/}"
		line="${line//[[:space:]]/}"

		[[ -z "$line" ]] && continue

		if ! __sdkman_matches_candidate_format "$line"; then
			__sdkman_echo_red "Invalid candidate format!"
			echo ""
			__sdkman_echo_yellow "Expected 'candidate=version' but found '$line'"

			return 1
		fi

		"$cb" "${line%=*}" "${line#*=}"
	done < "$sdkmanrc"
```

It's actually moire *in line* (get it 😄 ?) with how I would solve the problem in a high level language like Groovy or Kotlin

```kotlin
File(".sdkmanrc")
    .readLines()
    .map { ... }
    .forEach { ... }
```

The idiomatic way in Bash 4+ would be to use `readarray` / `mapfile` for this, which also reads the whole file into an array.
But this would break compatibility with Bash 3.

**Comment by helpermethod** [comment]
Hm, okay, `sdkman_auto_answer`may still make sense. Not fun being asked if every candidate should become the new default.

**Comment by helpermethod** [comment]
Made the necessary changes. Needs more testing, though.

**Comment by helpermethod** [comment]
How about we introduce another config value like `sdkman_set_default` with default `y`? Then I could overwrite it when calling `__sdk_install`? It would also allow to make `USE` a local variable, with `$sdkman_set_default` as its default value (instead of '').

**Comment by joehorsnell** [comment]
This is great @helpermethod. It would be very useful to get this PR ready and merged as without it, the `env` command is of limited use, IMO.

Need any help getting this PR over the line?

**Comment by helpermethod** [comment]
@joehorsnell I will pick up the issue again as soon as the `sdk env clear` command is merged.

**Comment by joehorsnell** [comment]
Hey @helpermethod, was #825 the only blocker for this PR?

**Comment by helpermethod** [comment]
@joehorsnell There was a regression in one of the last PRs but today I will continue working on the ticket! Stay tuned!

**Comment by helpermethod** [comment]
Implementation already works but needs more tests.

image

**Comment by helpermethod** [comment]
Closing in favor of #839.

**Comment by helpermethod** [comment]
Closing this in favor of #839.

**Comment by helpermethod** [comment]
Closing this in favor of #839.

**Pull request closed without merging**